hal/src/main/native/mau/DriveStation/MauDriveData.cpp

```diff
@@ -0,0 +1,229 @@
+#include "MauDriveData.h"
+#include <cstring>
+
+wpi::priority_mutex Mau_DriveData::memLock;
+wpi::priority_condition_variable Mau_DriveData::memSignal;
+
+HAL_AllianceStationID Mau_DriveData::allianceID;
+HAL_MatchInfo Mau_DriveData::matchInfo;
+char Mau_DriveData::infoEventName[Mau_kMatchNameLength];
+char Mau_DriveData::infoGameMessage[Mau_kMatchMessageLength];
+HAL_ControlWord Mau_DriveData::controlWord;
+Mau_SharedJoystick Mau_DriveData::joysticks[6];
+
+void Mau_DriveData::unlockAndSignal() {
+    memSignal.notify_all();
+    memLock.unlock();
+}
+
+
+ void Mau_DriveData::initializeDriveData() {
+	infoEventName[0] = 0;
+	matchInfo.eventName = infoEventName;
+	infoGameMessage[0] = 0;
+	matchInfo.gameSpecificMessage = infoGameMessage;
+ }
+
+void Mau_DriveData::updateAllianceID(HAL_AllianceStationID id) {
+    memLock.lock();
+    allianceID = id;
+    unlockAndSignal();
+}
+
+void Mau_DriveData::updateMatchInfo(HAL_MatchInfo* info) {
+    memLock.lock();
+    matchInfo.eventName = info->eventName;
+    matchInfo.matchType = info->matchType;
+    matchInfo.matchNumber = info->matchNumber;
+    matchInfo.replayNumber = info->replayNumber;
+    matchInfo.gameSpecificMessage = info->gameSpecificMessage;
+    unlockAndSignal();
+}
+
+void Mau_DriveData::updateMatchType(HAL_MatchType type) {
+    memLock.lock();
+    matchInfo.matchType = type;
+    unlockAndSignal();
+}
+
+// --- Update: ControlWord --- //
+
+void Mau_DriveData::updateIsEnabled(bool enabled) {
+    memLock.lock();
+    controlWord.enabled = enabled;
+    unlockAndSignal();
+}
+
+void Mau_DriveData::updateIsAutonomous(bool auton) {
+    memLock.lock();
+    controlWord.autonomous = auton;
+    unlockAndSignal();
+}
+
+void Mau_DriveData::updateIsTest(bool test) {
+    memLock.lock();
+    controlWord.test = test;
+    unlockAndSignal();
+}
+
+void Mau_DriveData::updateEStop(bool eStop) {
+    memLock.lock();
+    controlWord.eStop = eStop;
+    unlockAndSignal();
+}
+
+void Mau_DriveData::updateIsFmsAttached(bool fms) {
+    memLock.lock();
+    controlWord.fmsAttached = fms;
+    unlockAndSignal();
+}
+
+void Mau_DriveData::updateIsDsAttached(bool ds) {
+    memLock.lock();
+    controlWord.dsAttached = ds;
+    unlockAndSignal();
+}
+
+// --- Update: Joystick --- //
+
+void Mau_DriveData::updateJoyAxis(int joyNumber, int16_t axisCount, int8_t* axes) {
+    memLock.lock();
+    joysticks[joyNumber].joyAxes.count = axisCount;
+    for (int index = 0; index < axisCount; index++) {
+        joysticks[joyNumber].joyAxes.axes[index] = ((float)axes[index]) / 127;
```

Comment:
After review; it's required within Mau (this is a good place to do it, I think) that the value of any joystick axis cannot exceed 1.0 or fall below -1.0.  Currently, this code allows an axis value of -1.01 - so this needs to be fixed.